### PR TITLE
DBZ-1224 SqlServer Snapshot Select Overrides

### PR DESCRIPTION
**Pull request by rsatishm**
Based on PostgreSql connector existing settings, introduced below two settings into SqlServer Connector.
1. Introduced Snapshot Select Override setting.
2. Introduced rows.fetch.size configuration setting.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Hey @rsatishm, thanks a lot! We'll take a look into this asap; in the meantime, could you:

* File a JIRA issue for this feature request and prefix the commit messages with its key
* Use an email to author the commits that's [linked](https://help.github.com/en/articles/why-are-my-contributions-not-showing-up-on-my-profile) to your GitHub account?

Thanks!

**Comment by rsatishm** [comment]
> Hey @rsatishm, thanks a lot! We'll take a look into this asap; in the meantime, could you:
> 
> * File a JIRA issue for this feature request and prefix the commit messages with its key
> * Use an email to author the commits that's [linked](https://help.github.com/en/articles/why-are-my-contributions-not-showing-up-on-my-profile) to your GitHub account?
> 
> Thanks!

Thanks. I have filed the Jira issue - https://issues.jboss.org/browse/DBZ-1224
Also prefixed all my commit messages with the issue#

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -193,8 +199,13 @@ protected void complete(SnapshotContext snapshotContext) {
     }
 
     @Override
-    protected String getSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
-        return String.format("SELECT * FROM [%s].[%s]", tableId.schema(), tableId.table());
+    protected Optional<String> getSnapshotSelect(SnapshotContext snapshotContext, TableId tableId) {
+        if (snapshotOverrides.containsKey(tableId)) {
```
Comment (jpechane) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java]: This should also go up to base class. The base class should check if there are overrides and if yes then use them. If not then this method should be called.

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -362,13 +365,27 @@ private Timer getTableScanLogTimer() {
...
     // scn xyz")
-    protected abstract String getSnapshotSelect(SnapshotContext snapshotContext, TableId tableId);
+    /**
+     * Generate a valid postgres query string for the specified table, or an empty {@link Optional}
+     * to skip snapshotting this table (but that table will still be streamed from)
+     *
+     * @param tableId the table to generate a query for
+     * @return a valid query string, or none to skip snapshotting this table
+     */
+    protected abstract Optional<String>getSnapshotSelect(SnapshotContext snapshotContext, TableId tableId);
```
[the remark this answers is not in the document]
Comment (rsatishm) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: This method is now moved to base. It returns String, no more optional. Whether to skip a snapshot is done well before as we build the tables to be captured

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 3]
```diff
@@ -309,48 +311,49 @@ private void createDataEventsForTable(ChangeEventSourceContext sourceContext, Sn
...
-                            Strings.duration(stop - exportStart));
-                    snapshotProgressListener.rowsScanned(table.id(), rows);
-                    logTimer = getTableScanLogTimer();
-                }
-
-                dispatcher.dispatchSnapshotEvent(table.id(), getChangeRecordEmitter(snapshotContext, row),
-                        snapshotReceiver);
+                LOGGER.info("\t Finished exporting {} records for table '{}'; total duration '{}'", rows,
+                        table.id(), Strings.duration(clock.currentTimeInMillis() - exportStart));
+            } catch (SQLException e) {
```
[the remark this answers is not in the document]
Comment (rsatishm) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: Yes, now that the select doesn't return optional, there is nothing after '}'

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java [hunk 4]
```diff
@@ -203,6 +214,36 @@ protected ChangeRecordEmitter getChangeRecordEmitter(SnapshotContext snapshotCon
         return new SnapshotChangeRecordEmitter(snapshotContext.offset, row, getClock());
     }
 
+    @Override
```
[the remark this answers is not in the document]
Comment (rsatishm) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java]: Yes, did that now.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 5]
```diff
@@ -231,13 +231,34 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
                     + "When '" + SnapshotIsolationMode.SNAPSHOT.getValue() + "' is specified, connector runs the initial snapshot in SNAPSHOT isolation level, which guarantees snapshot consistency. In addition, neither table nor row-level locks are held. "
                     + "In '" + SnapshotIsolationMode.READ_UNCOMMITTED.getValue() + "' mode neither table nor row-level locks are acquired, but connector does not guarantee snapshot consistency.");
 
+    public static final Field ROWS_FETCH_SIZE = Field.create("rows.fetch.size")
```
[the remark this answers is not in the document]
Comment (rsatishm) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: Done

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java (hunk 1 above)
Comment (rsatishm) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java]: Done

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java [hunk 6]
```diff
@@ -203,6 +214,36 @@ protected ChangeRecordEmitter getChangeRecordEmitter(SnapshotContext snapshotCon
...
+    protected Statement readTableStatement() throws SQLException {
+        int rowsFetchSize = connectorConfig.rowsFetchSize();
+        Statement statement = jdbcConnection.connection().createStatement(); // the default cursor is FORWARD_ONLY
+        statement.setFetchSize(rowsFetchSize);
+        return statement;
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (rsatishm) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java]: Introduced SnapshotStatementFactory that mysql, postgres and sqlserver uses to create statement for snaposhot

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 7]
```diff
@@ -368,7 +382,21 @@ private Timer getTableScanLogTimer() {
...
     // TODO Handle override option generically; a problem will be how to handle the dynamic part (Oracle's "... as of
     // scn xyz")
-    protected abstract String getSnapshotSelect(SnapshotContext snapshotContext, TableId tableId);
+    /**
+     * Generate a valid sqlserver query string for the specified table
+     *
+     * @param tableId the table to generate a query for
+     * @return a valid query string
+     */
+    private String getSnapshotSelect(TableId tableId) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: This base class is shared between different implementations for SQL Server, Oracle etc, i.e. it cannot contain any specifics of SQL Server. For that reason it should remain an abstract method as before.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 8]
```diff
@@ -320,4 +335,20 @@ protected boolean isPositionAtOrBefore(Document recorded, Document desired) {
             }
         };
     }
+
+    /**
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: These two methods could go to the base class?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 9]
```diff
@@ -231,13 +231,27 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
                     + "When '" + SnapshotIsolationMode.SNAPSHOT.getValue() + "' is specified, connector runs the initial snapshot in SNAPSHOT isolation level, which guarantees snapshot consistency. In addition, neither table nor row-level locks are held. "
                     + "In '" + SnapshotIsolationMode.READ_UNCOMMITTED.getValue() + "' mode neither table nor row-level locks are acquired, but connector does not guarantee snapshot consistency.");
 
+    public static final Field ROWS_FETCH_SIZE = RelationalDatabaseConnectorConfig.ROWS_FETCH_SIZE.withDefault(DEFAULT_ROWS_FETCH_SIZE);
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: What's the reasoning for (re-)defining these fields here? They should be in the base class.

Comment (rsatishm) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: Postgres has the default rows fetch set to 10240 whereas sqlserver has the default was set to 2000. The base will have one of the default set, how do we override the default in the specific database config?

Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: I think we can converge on default value; just go for the larger one globally.

**Comment by gunnarmorling** [comment]
Hey @rsatishm, something seems to went wrong a bit here: there are way more commits in the PR than there should. Could you rebase the PR to the current master?

And could you also please use an e-mail associated to your GH account for authoring the commit; that way it will be [properly attributed](https://help.github.com/en/articles/why-are-my-contributions-not-showing-up-on-my-profile) to you in the project stats. 

Thanks!

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 10]
```diff
@@ -27,6 +27,8 @@
  */
 public abstract class RelationalDatabaseConnectorConfig extends CommonConnectorConfig {
 
+    private static final int DEFAULT_ROWS_FETCH_SIZE = 10240;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java]: This has been taken care of already as part of https://github.com/debezium/debezium/pull/855. The code here won't be needed any longer, it should become apparent during the rebase.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 11]
```diff
@@ -226,6 +226,8 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
             RelationalDatabaseConnectorConfig.SERVER_NAME,
             DATABASE_NAME,
             SNAPSHOT_MODE,
+            RelationalDatabaseConnectorConfig.ROWS_FETCH_SIZE,
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java]: Not needed any longer, see my other comment.

**Comment by rsatishm** [comment]
> Hey @rsatishm, something seems to went wrong a bit here: there are way more commits in the PR than there should. Could you rebase the PR to the current master?
> 
> And could you also please use an e-mail associated to your GH account for authoring the commit; that way it will be [properly attributed](https://help.github.com/en/articles/why-are-my-contributions-not-showing-up-on-my-profile) to you in the project stats.
> 
> Thanks!

I see now snapshot row fetch setting is part of common config so would that be the reason? Also now my other part of change which configures the row fetch size is no more required. I will rebase the PR to the current master.

**Comment by gunnarmorling** [comment]
> I see now snapshot row fetch setting is part of common config so would that be the reason?

I reckon at some point you did a merge instead of rebase. If you rebase it, the PR eventually should only contain your commit(s).

**Comment by gunnarmorling** [comment]
Hey @rsatishm, just a quick heads-up: we're planning to release Debezium 0.9.5 tomorrow, so if the things we discussed by then, we could include it. If not, it's no problem either, as we'll release every three weeks. So please don't feel pressured about it, just wanted to bring it up, just in case you had planned to wrap this up today anyways. Thanks!

**Comment by gunnarmorling** [comment]
Hey @rsatishm, howdy? How are things, are you still planning to wrap this one up?

**Comment by rsatishm** [comment]
> Hey @rsatishm, howdy? How are things, are you still planning to wrap this one up?

Hi @gunnarmorling, my apologies, for some reason my stuck with the git login issues. I am able to login here in github but the same credentials some doesn't work in my IDE environment. I couldn't figure out yet why the login credentials doesn't work.

**Comment by gunnarmorling** [comment]
Doh, sorry to hear that. So what should we do then, should someone else take over this one? Or perhaps you try with a new IDE set-up to get unstuck?

Would be great to come to closure with this one in one way or another.

**Comment by gunnarmorling** [comment]
I'll try and rebase this and completing any outstanding task related to this.

**Comment by gunnarmorling** [comment]
Superseded by https://github.com/debezium/debezium/pull/888. I'll close this one here in favor of the other one (which contains the exact same diff) for the sake of a clean commit history. I'll rebase that other one next now. Thanks, @rsatishm!

**Pull request closed without merging**